improvement(client-core-interfaces): revise JsonDeserialized recursion

jason-ha opened the pull request:
Use `OpaqueJsonDeserialized` at recursion points for inexactly deserializable types instead of limited unrolling. Callers will need to use cast helper to resolve, which will produce a repeatable pattern.

Supporting changes:
 1. Allow `OpaqueJson*` branded types (classes) when checking for recursion and hidden (non-public) properties.
 1. Separate `RecursionMarker` `AllowExactly` logic re-use from the `AllowExactly` use in `JsonDeserializedOpaqueConversion`
 1. Internally, ancestor types are now explicitly limited to `object` types.
 1. Remove unused `Control` option in `JsonSerializableImpl`'s use of `IfNonPublicProperties`.
 1. Indexed properties with unsupported types no longer gain `| undefined` by being explicitly optional. 
 1. Better handle branded index keys - TypeScript has an issue comparing indices that have been manipulated. Branded indices like `type FooBarString = string & { foo: "bar" }`, when filtered, may lose `FooBarString` and `string & { foo: "bar" }` will not match via `IfSameType` implementation.
    - Make sure to use original key type (when possible, looks like all cases), when generating keys for mapped types.
 1. Branded property keys are assumed to be indexed properties. (Complete test is not currently possible and it is impractical to use a branded literal property key.)
    - Introduce `IfIndexOrBrandedKey` and replace all `IfIndexKey` use.
    - Relocate `IfIndexKey` to common section and rename to `IfVariableStringOrNumber`.
 1. Tests:
    - Update case for class with private data in a recursive situation.
    - Add explicit verification that no result contains `any`.
    - Add branded index keys cases.
 1. Correct RawInboundExtensionMessage  to use OpaqueJsonDeserialized
 1. Correct Presence test case showing missing OpaqueJson wrapping

jason-ha committed:
improvement(client-core-interfaces): limit JsonDeserialized recursion to objects
Recursion can only happen with object types. Restrict to that and use `NonNullJsonObjectWith` to more accurately represent results.

jason-ha committed:
improvement(client-core-interfaces): revise JsonDeserialized recursion
Use `OpaqueJsonDeserialized` at recursion points for inexactly deserializable types instead of limited unrolling. Callers will need to use cast helper to resolve which will produce a repeatable pattern.

Update tests with case for class with private data in a recursive situation.

tylerbutler commented on packages/common/core-interfaces/src/exposedInternalUtilityTypes.ts:
```diff
@@ -1062,28 +1062,26 @@ export namespace InternalUtilityTypes {
-	 * @privateRemarks
-	 * This utility is reentrant and will process a type `T` up to `RecurseLimit`.
+	 * @remarks
+	 * If no modification is needed (`T` is exactly deserilizable), `T` will
+	 * be the result.
```
```suggestion
	 * If no modification is needed (`T` is exactly deserializable), `T` will
	 * be the result.
```

jason-ha committed:
merge: 'main' into core/opaque-json-for-imperfect-recursion

jason-ha committed:
docs(client-core-interfaces): typo correction

jason-ha committed:
improvement(client-container-runtime): use extension message type correctness

jason-ha committed:
docs(client-core-interfaces): "Opaque Json" -> "OpaqueJson*"

jason-ha committed:
improvement(client-core-interfaces): remove unused Control option

jason-ha committed:
fix(client-core-interfaces): basic OpaqueJson in recursion support
- Allow those branded types (classes) when checking for recursion and hidden (non-public) properties
- Separate `RecursionMarker` `AllowExactly` logic re-use from the `AllowExactly` use in `JsonDeserializedOpaqueConversion`

jason-ha committed:
test(client-core-interfaces): test for `any` results

jason-ha committed:
merge: 'main' into core/opaque-json-for-imperfect-recursion-WIP

jason-ha committed:
fix(client-core-interfaces): handle branded indices
TypeScript has an issue comparing indices that have been manipulated. Branded indices like `type FooBarString = string & { foo: "bar" }` when filtered may lose `FooBarString` and `string & { foo: "bar" }` will not match via `IfSameType` implementation.
=> Make sure to use original key type (when possible, looks like all cases), when generating keys for mapped type.

Additionally:
- update indexed key tests to consider all branded keys as indexed (as distinguishing literal branded from templated is currently impossible [generically] and there is no known practical use for branded literals in JSON context).
- correct Presence test case now shown as missing opaque wrapping.

jason-ha committed:
fix(client-core-interfaces): JsonDeserialized index types with unsupported values
now have true result limited to supported value types rather than `| undefined`.

There is no need to apply optional `?` to a partially supported index key-value as indexed keys are inherently optional. Check for that when processing properties.

Additionally relocate `IfIndexKey` to common section as used by both.

jason-ha committed:
merge: 'main' into core/opaque-json-for-imperfect-recursion

jason-ha committed:
test(client-core-interfaces): revealOpaqueJson recursion option

jason-ha committed:
merge: 'main' into core/opaque-json-for-imperfect-recursion
